File: c_files/palindrome.c/palindrome.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>

// Συνάρτηση για να βρει το μέγιστο μήκος μιας παλινδρομικής υποσυμβολοσειράς και την αρχική της θέση
int longest_palindromic_substring(char* s, int* max_length, int* start_index) {
    int n = strlen(s);
    if (n == 0) {
        return -1;
    }

    // Δημιουργία και αρχικοποίηση του πίνακα dp
    bool dp[n][n];
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            dp[i][j] = false;
        }
    }

    *max_length = 1;
    *start_index = 0;

    // Όλες οι συμβολοσειρές μήκους 1 είναι παλινδρομικές
    for (int i = 0; i < n; i++) {
        dp[i][i] = true;
    }

    // Έλεγχος για μεγαλύτερα μήκη
    for (int length = 2; length <= n; length++) {
        for (int i = 0; i < n - length + 1; i++) {
            int j = i + length - 1;
            if (s[i] == s[j] && dp[i + 1][j - 1]) {
                dp[i][j] = true;
                *start_index = i;
                *max_length = length;
            }
        }
    }
    return 0;
}

int main() {
    char s[] = "nisonanomimatamimonanosin";
    int max_length;
    int start_index;

    int returned = longest_palindromic_substring(s, &max_length, &start_index);
    if (returned < 0) {
        fprintf(stderr, "Fuck You.\n");
        exit(returned);
    }

    printf("Length of original string: %ld\n", strlen(s));
    printf("Max length of palindrome substring: %d\n", max_length);
    printf("Starting position of palindrome substring: %d\n", start_index+1);

    return 0;
}

File: python/palindrome/palindrome.py
def longest_palindromic_substring(s):
    n = len(s)
    if n == 0:
        return 0, -1

    # Δημιουργία του πίνακα dp και αρχικοποίηση
    dp = [[False] * n for _ in range(n)]

    max_length = 1
    start_index = 0

    # Όλες οι συμβολοσειρές μήκους 1 είναι παλινδρομικές
    for i in range(n):
        dp[i][i] = True

    # Έλεγχος για συμβολοσειρές μήκους 2
    for i in range(n - 1):
        if s[i] == s[i + 1]:
            dp[i][i + 1] = True
            start_index = i
            max_length = 2

    # Έλεγχος για μεγαλύτερα μήκη
    for length in range(3, n + 1):  # length είναι το μήκος της υποσυμβολοσειράς
        for i in range(n - length + 1):
            j = i + length - 1  # Το τέλος της υποσυμβολοσειράς
            if s[i] == s[j] and dp[i + 1][j - 1]:
                dp[i][j] = True
                start_index = i
                max_length = length

    return max_length, start_index

# Παράδειγμα χρήσης
s = "nisonanomimatamimonanosin"
max_length, start_index = longest_palindromic_substring(s)
print("Length of starting string:", len(s))
print("Maximum length of palindrome substring:", max_length)
print("Starting position of palindrome substring:", start_index+1)
